app/scripts/metamask-controller.js

```diff
@@ -643,7 +653,60 @@ export default class MetamaskController extends EventEmitter {
       messenger: networkControllerMessenger,
       state: initialNetworkControllerState,
       infuraProjectId: opts.infuraProjectId,
+      getRpcServiceOptions: () => ({
+        fetch: globalThis.fetch.bind(globalThis),
+        btoa: globalThis.btoa.bind(globalThis),
+      }),
     });
+    networkControllerMessenger.subscribe(
+      'NetworkController:rpcEndpointUnavailable',
+      async ({ chainId, endpointUrl, error }) => {
+        const parsedUrl = new URL(endpointUrl);
+        const hostParts = parsedUrl.host.split('.');
+        const rootDomainName = hostParts.slice(-2).join('.');
+        if (
+          (rootDomainName === 'infura.io' ||
+            rootDomainName === 'quicknode.pro') &&
+          isConnectionError(error)
+        ) {
+          console.log('RPC endpoint unavailable', {
+            chainId,
+            endpointUrl,
+            error,
+          });
+          this.metaMetricsController.trackEvent({
+            category: MetaMetricsEventCategory.Network,
+            event: MetaMetricsEventName.RpcServiceUnavailable,
+            properties: {
+              chain_id_caip: `eip155:${chainId}`,
+              rpc_endpoint_url: onlyKeepHost(endpointUrl),
+            },
+          });
+        }
+      },
+    );
+    networkControllerMessenger.subscribe(
+      'NetworkController:rpcEndpointDegraded',
+      async ({ chainId, endpointUrl }) => {
+        const parsedUrl = new URL(endpointUrl);
+        const hostParts = parsedUrl.host.split('.');
+        const rootDomainName = hostParts.slice(-2).join('.');
+        if (
+          rootDomainName === 'infura.io' ||
+          rootDomainName === 'quicknode.pro'
+        ) {
+          console.log('RPC endpoint degraded', { chainId, endpointUrl });
+          this.metaMetricsController.trackEvent({
+            category: MetaMetricsEventCategory.Network,
+            event: MetaMetricsEventName.RpcServiceDegraded,
+            properties: {
+              chain_id_caip: `eip155:${chainId}`,
+              rpc_endpoint_url: endpointUrl,
+            },
+          });
+        }
+      },
+    );
```

Comment:
Can you move these to a TypeScript utility file? We're trying to write less JS and more TS 😅

Reply:
Do you want me to move the subscriptions to a utility file, or extract some kind of `getRootDomainName` function to a utility file?

Reply:
Done here: [6143cc0](https://github.com/MetaMask/metamask-extension/pull/31089/commits/6143cc064221fe568bca7d9f7e7c6324ef42088b)